src/controllers/controllerengine.cpp

```diff
@@ -98,30 +98,44 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
     }
 }
 
-/* -------- ------------------------------------------------------
-Purpose: Resolves a function name to a QScriptValue including
-            OBJECT.Function calls
-Input:   -
-Output:  -
--------- ------------------------------------------------------ */
-QScriptValue ControllerEngine::resolveFunction(const QString& function) const {
+/* ------------------------------------------------------------------
+Purpose: Turn a snippet of JS into a QScriptValue function.
+         Wrapping it in an anonymous function allows any JS that
+         evaluates to a function to be used in MIDI mapping XML files
+         and ensures the function is executed with the correct
+         'this' object.
+Input:   QString snippet of JS that evaluates to a function
+Output:  QScriptValue of JS snippet wrapped in an anonymous function
+------------------------------------------------------------------- */
+QScriptValue ControllerEngine::wrapFunctionCode(const QString& codeSnippet) {
+    QScriptValue wrappedFunction;
+
     QHash<QString, QScriptValue>::const_iterator i =
-            m_scriptValueCache.find(function);
-    if (i != m_scriptValueCache.end()) {
-        return i.value();
-    }
+            m_scriptWrappedFunctionCache.find(codeSnippet);
 
-    QScriptValue object = m_pEngine->globalObject();
-    QStringList parts = function.split(".");
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
+            return m_pEngine->evaluate("(function () {})");
+        }
+        if (checkException()) {
+            return m_pEngine->evaluate("(function () {})");
+        }
 
-    for (int i = 0; i < parts.size(); i++) {
-        object = object.property(parts.at(i));
-        if (!object.isValid()) {
-            break;
+        QString wrappedCode = "(function (channel, control, value, status, group) { (" +
+                              codeSnippet + ")(channel, control, value, status, group); })";
+        wrappedFunction = m_pEngine->evaluate(wrappedCode);
+
+        if (checkException()) {
+            // There will be an exception if the codeSnippet didn't evaluate to a function
+            return m_pEngine->evaluate("(function () {})");
         }
+
+        m_scriptWrappedFunctionCache[codeSnippet] = wrappedFunction;
```

Comment:
Hm, are you sure that error occurred in evaluating the wrapper function or was it executing the wrapper function? 

I wrote a quick test that suggests evaluating the wrapper would work fine in that case:
https://github.com/rryan/mixxx/commit/7036720de8064e273e0ab0dd9286cf19586c21c5

Output:
```
[ RUN      ] ControllerEngineTest.Evaluate
Loading resources from  "/Users/rjryan/Code/mixxx/res/"
ConfigObject: Could not read "/Users/rjryan/Code/mixxx/src/test/test_data/test.cfg"
SetUp
ControllerEngine: Loading "/var/folders/qr/klrx2z4s0l37vkk8y5lq5718002y8v/T/qt_temp.M56742"
TypeError: Result of expression '(P32.whatever)' [stuff] is not a function.
whatever works!
TearDown
ControllerEngine shutting down...
Warning: Test leaked control: "[Test]" "potmeter_toggle"
Warning: Test leaked control: "[Test]" "potmeter_up_small"
Warning: Test leaked control: "[Test]" "potmeter_set_default"
Warning: Test leaked control: "[Test]" "potmeter_down_small"
Warning: Test leaked control: "[Test]" "potmeter_set_one"
Warning: Test leaked control: "[Test]" "potmeter_set_minus_one"
Warning: Test leaked control: "[Test]" "potmeter_down"
Warning: Test leaked control: "[Test]" "potmeter_minus_toggle"
Warning: Test leaked control: "[Test]" "potmeter_set_zero"
Warning: Test leaked control: "[Test]" "potmeter"
Warning: Test leaked control: "[Test]" "potmeter_up"
[       OK ] ControllerEngineTest.Evaluate (1 ms)

```